### PR TITLE
ENH: add variable bits to `stats.qmc.Sobol`

### PR DESCRIPTION
**Pull request by tupui**
Linked to #14510 (enhancement list) and might help with #13474 (performance).

Allows to change the number of `bits`. It's linked to the dtype, but one can have any integer value from 0 to 64.

Internally, the bits value represents the number of possible values and the precision. Sobol' sequence produce integers with a value from 0 to `2**bits` and the value is scaled by `2**bits`. Before this PR, we had `bits=30` and then we could only generate `2**30` points. With this PR, you can now have from `2**0=1` to `2**64` points.

For this to work, the types of internals need to be able to hold large enough integers (again up to `2**bits`). Previously, most types were set to `int`. For now I changed all types (just the ones I thought needed to be changed) to `cnp.uint64_t`.

Mostly ready, I still want to check if having a variable dtype could help reduce the computational cost/memory. Internal matrices are of size `dim*bits`. For high dimensions, this might matter.

Among the checks I added for `bits`, I also added a check to forbid sampling past `2**bits` with a nice user facing message. It's technically a breaking change, but this would result in extremely wrong values not even in the [0, 1] interval.

### CODE REVIEW

**Review by tirthasheshpatel — COMMENTED** [review]
A few comments on possible overflows. This led to type promotion of numpy arrays from `uint64` to `object`:

```py
>>> from scipy.stats import qmc
>>> rng = qmc.Sobol(d=1, bits=64)
>>> rng.random()
array([[0.5866267744320944]], dtype=object)
```

File: scipy/stats/_sobol.pyx [hunk 1]
```diff
@@ -232,40 +229,49 @@ cpdef void initialize_v(cnp.int_t[:, :] v, const int dim):
...
+                 const int maxbit,
+                 cnp.uint64_t[:, :] sv,
+                 cnp.uint64_t[:] quasi,
+                 cnp.float_t[:, :] sample) nogil:
+    cdef int j, l
+    cdef cnp.uint64_t num_gen_loc = num_gen
+    cdef cnp.uint64_t i, qtmp
+
+    # largest possible integer
+    cdef cnp.uint64_t largest_number = 2 ** maxbit
```
[the remark this answers is not in the document]
Comment (tupui) [reply scipy/stats/_sobol.pyx:252]: For the scale I am not sure because in his code, Sergei Kucherenko is just checking to not return more than `2**maxbit-1` points, but has `2**maxbit` for the scale. I will send him an email to check these.

File: scipy/stats/_sobol.pyx [hunk 2]
```diff
@@ -289,24 +295,26 @@ cdef int cdot_pow2(cnp.int_t[:] a) nogil:
 @cython.boundscheck(False)
 @cython.wraparound(False)
 cpdef void _cscramble(const int dim,
-                      cnp.int_t[:, :, :] ltm,
-                      cnp.int_t[:, :] sv) nogil:
-    cdef int d, i, j, k, l, lsm, lsmdp, p, t1, t2, vdj
+                      const int maxbit,
+                      cnp.uint64_t[:, :, :] ltm,
+                      cnp.uint64_t[:, :] sv) nogil:
+    cdef int d, i, j, k, l, p
```
[the remark this answers is not in the document]
Comment (tupui) [reply scipy/stats/_sobol.pyx:301]: Indeed! You can make a suggestion so that we have you here.

**Review by tupui — COMMENTED** [review]
Thanks @tirthasheshpatel! I will let you know about the max value. Something else I am seeing that could improve performance is to mark contiguous arrays (`[::1]`).

If you see things, feel free to commit on the branch.

File: scipy/stats/_sobol.pyx [hunk 3]
```diff
@@ -289,24 +295,26 @@ cdef int cdot_pow2(cnp.int_t[:] a) nogil:
...
+    cdef cnp.uint64_t lsmdp, t1, t2, vdj
```
Comment (tirthasheshpatel) [inline scipy/stats/_sobol.pyx:302]:
```suggestion
    cdef int d, i, j, k, p
    cdef cnp.uint64_t l, lsmdp, t1, t2, vdj
```

File: scipy/stats/_sobol.pyx (hunk 2 above)
[the remark this answers is not in the document]
Comment (tirthasheshpatel) [reply scipy/stats/_sobol.pyx:301]: Done.

**Review by tupui — COMMENTED** [review]
@tirthasheshpatel I had another read at the codes and I am pretty sure now (no replies yet) that scale needs to be 2^bits. But the way they are doing in other codes is that they pre-compute the value to prevent overflow. Hence I am now doing this in Python and passing the scale down. This prevents any type promotion to object. I added a test for 64bits too.

Also I added a benchmark, contiguous notation (did not see any change in the benchmark, but maybe I am not doing it correctly) and I tested to change the type and even with uint16 I see no difference with uint64 in terms of speed. So maybe we leave this out (or you have tricks to try and could see a difference).

Let me know what you think with the new adjustments 😃 (I've seen the MyPy issues, I will squash the change with the previous commit and force push, waiting for CI to finish before).

**Comment by tupui** [comment]
@mdhaber @chrisb83 this might be of interest 😃

**Comment by tupui** [comment]
Some additional elements on this PR:

* change `MAXBIT` to `maxbit` in Cython, it’s because now it’s a parameter instead of a constant. `RECIPD`, I don’t know what this originally meant so I just put a descriptive name which is also what other code tend to use.
* Added some small doc around.
* The underscore (`initialize_v` to `_ initialize_v`) is for some reason making the Cython code better supported by PyCharm. Not sure why, but this way it properly follow the links from the functions to the Cython. Same goes with using a relative import.
* `::1` indicates that the array is contiguous in memory following this axis. I could not measure any difference though. Maybe I am missing something. In any case a misuse would crash at compilation.
* In the benchmarks, I just tested the dimension and number of samples because the other parameters would have no impact. (scrambling just at init and the matrix `self._sv` is small enough that bits is also not impacting much)
* There is no easy path to go above 64 bits and from some discussions, no known use case.
* `self._quasi = self._shift.copy()` and `self.num_generated = 0` could be removed because they are respectively already done in the init and super init.

**Comment by mdhaber** [comment]
Thanks for posting that @tupui. I'll review what I can, but I have rarely used Cython, so mostly I'll have to trust that Cython-specific logic is OK.

File: scipy/stats/_qmc.py [hunk 4]
```diff
@@ -1319,47 +1323,56 @@ class Sobol(QMCEngine):
...
     ) -> None:
         super().__init__(d=d, seed=seed)
         if d > self.MAXDIM:
             raise ValueError(
-                "Maximum supported dimensionality is {}.".format(self.MAXDIM)
+                f"Maximum supported dimensionality is {self.MAXDIM}."
             )
 
+        self.MAXBIT = bits
+        if self.MAXBIT > 64:
```
Comment (mdhaber) [inline scipy/stats/_qmc.py:1338]: `MAXBIT` sounds like it is always going to be 64. Why is this not just `self.bits`, and the argument to `_initialize_v` not just `bits`?

File: scipy/stats/_qmc.py [hunk 5]
```diff
@@ -1319,47 +1323,56 @@ class Sobol(QMCEngine):
...
-            self, d: IntNumber, *, scramble: bool = True,
-            seed: SeedType = None
+        self, d: IntNumber, *, scramble: bool = True,
+        bits: IntNumber = 30, seed: SeedType = None
     ) -> None:
         super().__init__(d=d, seed=seed)
         if d > self.MAXDIM:
             raise ValueError(
-                "Maximum supported dimensionality is {}.".format(self.MAXDIM)
+                f"Maximum supported dimensionality is {self.MAXDIM}."
```
Comment (mdhaber) [inline scipy/stats/_qmc.py:1334]: +1 for f-strings (although a separate cleanup PR would be preferred.)

File: scipy/stats/_qmc.py [hunk 6]
```diff
@@ -1319,47 +1323,56 @@ class Sobol(QMCEngine):
...
+        self.MAXN = 2**self.MAXBIT
+
         # initialize direction numbers
-        initialize_direction_numbers()
+        _initialize_direction_numbers()
 
         # v is d x MAXBIT matrix
-        self._sv = np.zeros((d, self.MAXBIT), dtype=int)
-        initialize_v(self._sv, d)
+        self._sv = np.zeros((d, self.MAXBIT), dtype=np.uint64)
```
Comment (mdhaber) [inline scipy/stats/_qmc.py:1346]: I understand that if `dtype=int` this would be 32-bit, so going to `uint64` makes sense, and good to make it unsigned.

File: scipy/stats/_qmc.py [hunk 7]
```diff
@@ -1319,47 +1323,56 @@ class Sobol(QMCEngine):
...
             )
 
+        self.MAXBIT = bits
+        if self.MAXBIT > 64:
+            raise ValueError("Maximum supported 'bits' is 64")
+        self.MAXN = 2**self.MAXBIT
+
         # initialize direction numbers
-        initialize_direction_numbers()
+        _initialize_direction_numbers()
```
Comment (mdhaber) [inline scipy/stats/_qmc.py:1343]: I understand that "The underscore is for some reason making the Cython code better supported by PyCharm." This is fine (but again, if this is orthogonal to adding `bits`, I'd suggest it be in a separate PR).

File: scipy/stats/_qmc.py [hunk 8]
```diff
@@ -1319,47 +1323,56 @@ class Sobol(QMCEngine):
...
         else:
+            # scramble self._shift and self._sv
             self._scramble()
 
         self._quasi = self._shift.copy()
-        self._first_point = (self._quasi / 2 ** self.MAXBIT).reshape(1, -1)
+
+        # normalization constant with the largest possible number
+        # calculate in Python to not overflow int with 2**64
+        self._scale = 1.0 / 2 ** self.MAXBIT
```
Comment (mdhaber) [inline scipy/stats/_qmc.py:1359]: Giving a name to `_scale` seems to make the code a little simpler, OK.

**Review by mdhaber — COMMENTED** [review]
Just a start. I'll continue later. Some comments are just to confirm that I looked at the lines. Few require action; feel free to resolve yourself everything except the question about the name `MAXBIT` - I'd like to see the response.

**Review by tupui — COMMENTED** [review]
Thanks @mdhaber for having a look. FYI, @tirthasheshpatel has looked at the Cython part now.

For n-bits, I am still a bit experimenting. I reached out to my contacts to see how valuable it is to change the value itself. The initial intent was that I could then use another type to improve performance (speed and memory). Speed is unchanged, and the memory varies a bit. I am not sure how to properly benchmark the memory with ASV, if you have any experience here. I used `peakmem_xxx`. (I can also use `memory_profiler` if needed).

What do you think about the differences? Is it worth it? If not and the case of variable n-bits is not made from a quality/convergence perspective, then we could fix the value to 64 (note that this would be a backward compatibility break, which in such case is more than fine as we can argue the previous state was "slightly broken"). We can also have a private class attribute to still allow to change the value but not expose it on the API (until I have more insights on the performances, my preferred solution).

I also made some tests with changing the output from float64 to float32 and the results are similar. What we could do then is to just have 2 possibilities 32 or 64 bits. This might make more sense.


<details>
<summary>Memory benchmarks</summary>

```
# with uint64

			  ===== ======= ======= ======= =======
              --                 base2             
              ----- -------------------------------
                d      3       10      11      12  
              ===== ======= ======= ======= =======
                1    80.8M    81M    81.5M   81.5M 
                50   80.2M   81.1M   81.6M   79.2M 
               100    81M    80.7M    82M     85M  
              ===== ======= ======= ======= =======

# with uint16

              ===== ======= ======= ======= =======
              --                 base2             
              ----- -------------------------------
                d      3       10      11      12  
              ===== ======= ======= ======= =======
                1    76.8M   74.9M   75.4M   76.7M 
                50   75.6M   76.3M   75.4M   82.2M 
               100   76.8M   73.6M   77.7M   78.7M 
              ===== ======= ======= ======= =======
```

</details>


<details>
<summary>Convergence integration performance</summary>

`Sobol' scrambled` is here the current default which is 30bits. I did the exercise for a few functions of varying complexity and results are similar. Starting with 30 bits, results are similar to 64bits.

sobol_conv_integration_mean_Art 1


</details>

**Comment by tupui** [comment]
Had a chat with @mdhaber. Thanks again!
I am fine with this being merged as it is as we still have plenty of time to adjust things before a release.

@tirthasheshpatel @mdhaber if you could tell me what you prefer:

1. go with the current approach. Issues being that I am not convinced from my tests that having variable bits is useful
2. go with `bits`, `dtype` which can only be 32 or 64bits. I would use fused types so we have a 32 and 64 bits compiled code. And this would make a small backward incompatibility (32 vs 30 bits). As explained above, the previous behaviour can be considered a tiny bug and fine with me to break this.

If it's 1, feel free to go ahead with either more review or a merge. If 2, which terminology do you prefer?

2.1 `bits` only 32 or 64, I will adapt the input validation.
2.2 `dtype` either `np.float32` or `np.float64` this is more like a direct control on the output. I would add some doc to explain that it does not just change the shape of the output but control some internals.

As of now, my preference goes for 2.1.

**Comment by mdhaber** [comment]
>  small backward incompatibility (32 vs 30 bits)

So the exact numbers will change slightly (order ~1/2^30) or does the sequence totally change?

**Comment by mdhaber** [comment]
> So yeah, people might rely on this already and we would break them. If we don't want this, the current PR is good.

I think we need to maintain the 30-bit version, then.

One possibility is to expose `dtype` as the parameter with default `None`, which would maintain the old behavior. Explicit `np.float32` or `np.float64` would trigger one of the new behaviors.

**Comment by tirthasheshpatel** [comment]
I agree with @mdhaber. I like this PR as-is but I am not an expert here :) @tupui If you feel like only 32 and 64 bits are useful, I'd say go ahead with @mdhaber's suggestion.

**Comment by tupui** [comment]
> > So yeah, people might rely on this already and we would break them. If we don't want this, the current PR is good.
> 
> I think we need to maintain the 30-bit version, then.
> 
> One possibility is to expose `dtype` as the parameter with default `None`, which would maintain the old behavior. Explicit `np.float32` or `np.float64` would trigger one of the new behaviors. Or `bits` 32/64 as you suggested, but then I would not change the actual dtype.

Oh nice idea. In the end it's a technical point that common user do not need to know/deal with. Do you prefer another PR or I do this here? If we default to None, we could add a deprecation warning.

**Comment by mdhaber** [comment]
I still haven't reviewed enough to merge this one. I'll leave that to @tirthasheshpatel if he's close. Green CI couldn't hurt, of course.

But if you do merge this as-is, I could look at the followup with the `dtype` parameter. If it's not cumbersome to keep, don't know that we need to deprecate the old behavior.

**Comment by tupui** [comment]
CI failure is not related.

**Comment by mdhaber** [comment]
But you promised green CI : )

File: scipy/stats/_sobol.pyx [hunk 9]
```diff
@@ -93,14 +88,15 @@ def initialize_direction_numbers():
         np.savez_compressed("./_sobol_direction_numbers", vinit=vs, poly=poly)
 
     """
-    cdef int[:] dns_poly
-    cdef int[:, :] dns_vinit
+    cdef cnp.uint64_t[::1] dns_poly
+    cdef cnp.uint64_t[:, :] dns_vinit
```
Comment (mdhaber) [inline scipy/stats/_sobol.pyx:92]: I'll trust this is an appropriate data type and the change from `[:]` to `[::1]` is helpful.

File: scipy/stats/_sobol.pyx [hunk 10]
```diff
@@ -195,14 +191,18 @@ cdef int ibits(const int x, const int pos, const int length) nogil:
 
 @cython.boundscheck(False)
 @cython.wraparound(False)
-cpdef void initialize_v(cnp.int_t[:, :] v, const int dim):
-    cdef int d, i, j, k, m, p, newv, pow2
+cpdef void _initialize_v(
+    cnp.uint64_t[:, ::1] v, const int dim, const int maxbit
```
Comment (mdhaber) [inline scipy/stats/_sobol.pyx:195]: Again, I'll trust this. But if you don't see a performance difference, I don't see why to change it, especially here.

File: scipy/stats/_sobol.pyx [hunk 11]
```diff
@@ -289,24 +293,27 @@ cdef int cdot_pow2(cnp.int_t[:] a) nogil:
 @cython.boundscheck(False)
 @cython.wraparound(False)
 cpdef void _cscramble(const int dim,
-                      cnp.int_t[:, :, :] ltm,
-                      cnp.int_t[:, :] sv) nogil:
-    cdef int d, i, j, k, l, lsm, lsmdp, p, t1, t2, vdj
```
Comment (mdhaber) [inline scipy/stats/_sobol.pyx:294]: Was `lsm` not used?

File: scipy/stats/_qmc.py [hunk 12]
```diff
@@ -1319,47 +1323,56 @@ class Sobol(QMCEngine):
...
+        bits: IntNumber = 30, seed: SeedType = None
     ) -> None:
         super().__init__(d=d, seed=seed)
         if d > self.MAXDIM:
             raise ValueError(
-                "Maximum supported dimensionality is {}.".format(self.MAXDIM)
+                f"Maximum supported dimensionality is {self.MAXDIM}."
             )
 
+        self.maxbit = bits
```
[the remark this answers is not in the document]
Comment (tupui) [reply scipy/stats/_qmc.py:1337]: Yes my bad, I can call it `bits`.

File: scipy/stats/_qmc.py [hunk 13]
```diff
@@ -1377,21 +1390,41 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
         """
         sample = np.empty((n, self.d), dtype=float)
 
+        if n == 0:
+            return sample
```
[the remark this answers is not in the document]
Comment (tupui) [reply scipy/stats/_qmc.py:1394]: Actually it is needed because the `_draw` method would be called with `n-1=-1` which would overflow due to the change from `int` to `uint` (`OverflowError: can't convert negative value to npy_uint64`). It was working before but was accidental. Going into the function and doing nothing.

File: scipy/stats/_qmc.py [hunk 14]
```diff
@@ -1377,21 +1390,41 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
...
+        total_n = self.num_generated + n
+        if total_n > self.maxn:
+            msg = (
+                f"At most 2**{self.maxbit}={self.maxn} points can be "
+                f"generated. {self.num_generated} points have been previously "
+                f"generated, then: n={self.num_generated}+{n}={total_n}. "
+            )
+            if self.maxbit != 64:
+                msg += "Try to increase the 'bits' size."
+            raise ValueError(msg)
```
[the remark this answers is not in the document]
Comment (tupui) [reply scipy/stats/_qmc.py:1405]: We did not raise any error on this before no. You could ask more than 2^32 points and it would go on and give you non-sense.

File: scipy/stats/_sobol.pyx (hunk 10 above)
Comment (tupui) [reply scipy/stats/_sobol.pyx:195]: On my platform I don't see any change of performance, but I have a M1 which has a rather different memory architecture. Maybe someone with a "classical" architecture would see a benefit. From Cython doc, they recommend to do this and it should bring around 10% performance improvements.

@tirthasheshpatel proposed to check on his side.

File: scipy/stats/_sobol.pyx (hunk 11 above)
Comment (tupui) [reply scipy/stats/_sobol.pyx:294]: It was not no.

**Comment by tupui** [comment]
FYI I started the follow up work here https://github.com/tupui/scipy/pull/4

**Comment by mdhaber** [comment]
OK I'll go ahead and merge this. Doesn't seem to break anything.

File: scipy/stats/_sobol.pyx (hunk 10 above)
Comment (tirthasheshpatel) [reply scipy/stats/_sobol.pyx]: I did some experiments. Contiguous memoryviews don't change anything other than how indexing is done in C. So, I wrote a simple Cython extension with a non-contiguous memoryview:

```cython
import numpy as np
cimport numpy as np

np.import_array()

cdef np.uint64_t [:] my_arr
my_arr = np.zeros(5)

print(my_arr[1])
```

The generated C file uses the strides to index elements of the array:

```c
__pyx_v_12memory_views_my_arr.data + __pyx_t_5 * __pyx_v_12memory_views_my_arr.strides[0]
```

While, when I change the line `cdef np.uint64_t [:] my_arr` to `cdef np.uint64_t [::1] my_arr`, the only change is:

```c
__pyx_v_12memory_views_my_arr.data) + __pyx_t_5
```

i.e. strides are no longer used when indexing. If you have a large for loop, this could speed up things by 2-5% but doesn't give a great benefit.

The same is applicable to multidimensional array (only different being the speedup is even smaller)

<details>

<summary>Diff when memoryview is changed to contiguous memoryview in my little example</summary>

Cython version 0.29.28
NumPy version 1.22.2

```diff
diff --git a/cython_stuff/memory_views.c b/cython_stuff/memory_views.c
index df9b7d3..a435437 100644
--- a/cython_stuff/memory_views.c
+++ b/cython_stuff/memory_views.c
@@ -2024,7 +2024,7 @@ static int __Pyx_ValidateAndInit_memviewslice(
                 PyObject *original_obj);
 
 /* ObjectToMemviewSlice.proto */
-static CYTHON_INLINE __Pyx_memviewslice __Pyx_PyObject_to_MemoryviewSlice_ds_nn___pyx_t_5numpy_uint64_t(PyObject *, int writable_flag);
+static CYTHON_INLINE __Pyx_memviewslice __Pyx_PyObject_to_MemoryviewSlice_dc_nn___pyx_t_5numpy_uint64_t(PyObject *, int writable_flag);
 
 /* GCCDiagnostics.proto */
 #if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6))
@@ -17676,7 +17676,7 @@ static CYTHON_SMALL_CODE int __Pyx_InitCachedConstants(void) {
 
   /* "memory_views.pyx":7
  * 
- * cdef np.uint64_t [:] my_arr
+ * cdef np.uint64_t [::1] my_arr
  * my_arr = np.zeros(5)             # <<<<<<<<<<<<<<
  * 
  * print(my_arr[1])
@@ -18166,13 +18166,13 @@ if (!__Pyx_RefNanny) {
  * 
  * np.import_array()             # <<<<<<<<<<<<<<
  * 
- * cdef np.uint64_t [:] my_arr
+ * cdef np.uint64_t [::1] my_arr
  */
   __pyx_t_2 = __pyx_f_5numpy_import_array(); if (unlikely(__pyx_t_2 == ((int)-1))) __PYX_ERR(0, 4, __pyx_L1_error)
 
   /* "memory_views.pyx":7
  * 
- * cdef np.uint64_t [:] my_arr
+ * cdef np.uint64_t [::1] my_arr
  * my_arr = np.zeros(5)             # <<<<<<<<<<<<<<
  * 
  * print(my_arr[1])
@@ -18185,7 +18185,7 @@ if (!__Pyx_RefNanny) {
   __pyx_t_1 = __Pyx_PyObject_Call(__pyx_t_3, __pyx_tuple__21, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 7, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
-  __pyx_t_4 = __Pyx_PyObject_to_MemoryviewSlice_ds_nn___pyx_t_5numpy_uint64_t(__pyx_t_1, PyBUF_WRITABLE); if (unlikely(!__pyx_t_4.memview)) __PYX_ERR(0, 7, __pyx_L1_error)
+  __pyx_t_4 = __Pyx_PyObject_to_MemoryviewSlice_dc_nn___pyx_t_5numpy_uint64_t(__pyx_t_1, PyBUF_WRITABLE); if (unlikely(!__pyx_t_4.memview)) __PYX_ERR(0, 7, __pyx_L1_error)
   __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
   __PYX_XDEC_MEMVIEW(&__pyx_v_12memory_views_my_arr, 1);
   __pyx_v_12memory_views_my_arr = __pyx_t_4;
@@ -18208,7 +18208,7 @@ if (!__Pyx_RefNanny) {
     __Pyx_RaiseBufferIndexError(__pyx_t_2);
     __PYX_ERR(0, 9, __pyx_L1_error)
   }
-  __pyx_t_1 = __Pyx_PyInt_From_npy_uint64((*((__pyx_t_5numpy_uint64_t *) ( /* dim=0 */ (__pyx_v_12memory_views_my_arr.data + __pyx_t_5 * __pyx_v_12memory_views_my_arr.strides[0]) )))); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 9, __pyx_L1_error)
+  __pyx_t_1 = __Pyx_PyInt_From_npy_uint64((*((__pyx_t_5numpy_uint64_t *) ( /* dim=0 */ ((char *) (((__pyx_t_5numpy_uint64_t *) __pyx_v_12memory_views_my_arr.data) + __pyx_t_5)) )))); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 9, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_1);
   __pyx_t_3 = __Pyx_PyObject_CallOneArg(__pyx_builtin_print, __pyx_t_1); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 9, __pyx_L1_error)
   __Pyx_GOTREF(__pyx_t_3);
@@ -21412,17 +21412,17 @@ no_fail:
 }
 
 /* ObjectToMemviewSlice */
-  static CYTHON_INLINE __Pyx_memviewslice __Pyx_PyObject_to_MemoryviewSlice_ds_nn___pyx_t_5numpy_uint64_t(PyObject *obj, int writable_flag) {
+  static CYTHON_INLINE __Pyx_memviewslice __Pyx_PyObject_to_MemoryviewSlice_dc_nn___pyx_t_5numpy_uint64_t(PyObject *obj, int writable_flag) {
     __Pyx_memviewslice result = { 0, 0, { 0 }, { 0 }, { 0 } };
     __Pyx_BufFmt_StackElem stack[1];
-    int axes_specs[] = { (__Pyx_MEMVIEW_DIRECT | __Pyx_MEMVIEW_STRIDED) };
+    int axes_specs[] = { (__Pyx_MEMVIEW_DIRECT | __Pyx_MEMVIEW_CONTIG) };
     int retcode;
     if (obj == Py_None) {
         result.memview = (struct __pyx_memoryview_obj *) Py_None;
         return result;
     }
-    retcode = __Pyx_ValidateAndInit_memviewslice(axes_specs, 0,
-                                                 PyBUF_RECORDS_RO | writable_flag, 1,
+    retcode = __Pyx_ValidateAndInit_memviewslice(axes_specs, __Pyx_IS_C_CONTIG,
+                                                 (PyBUF_C_CONTIGUOUS | PyBUF_FORMAT) | writable_flag, 1,
                                                  &__Pyx_TypeInfo_nn___pyx_t_5numpy_uint64_t, stack,
                                                  &result, obj);
     if (unlikely(retcode == -1))
diff --git a/cython_stuff/memory_views.pyx b/cython_stuff/memory_views.pyx
index cd14a99..8ac389d 100644
--- a/cython_stuff/memory_views.pyx
+++ b/cython_stuff/memory_views.pyx
@@ -3,7 +3,7 @@ cimport numpy as np
 
 np.import_array()
 
-cdef np.uint64_t [:] my_arr
+cdef np.uint64_t [::1] my_arr
 my_arr = np.zeros(5)
 
 print(my_arr[1])
```

</details>

Comment (tupui) [reply scipy/stats/_sobol.pyx]: Thanks @tirthasheshpatel for the analysis. Well then if there is no real drawback, 2-5% in some cases is good to have.

**Comment by tupui** [comment]
Thank you both for your help and review! I opened the follow up we discussed #15712
